src/libraries/System.Private.CoreLib/src/System/Runtime/InteropServices/GCHandle.T.cs

```diff
@@ -0,0 +1,114 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Diagnostics.CodeAnalysis;
+using System.Runtime.CompilerServices;
+
+namespace System.Runtime.InteropServices
+{
+    /// <summary>
+    /// Represents an strongly-typed opaque, GC handle to a managed object.
+    /// A GC handle is used when an object reference must be reachable from
+    /// unmanaged memory.
+    /// </summary>
+    /// <remarks>
+    /// <para>This type corresponds to <see cref="GCHandleType.Normal"/>.</para>
+    /// <para>
+    /// This type is unsafe if used incorrectly. Incorrect usage like mismanagement
+    /// of lifetime, providing invalid handle value or concurrent disposal can result
+    /// into hard to diagnose crashes or data corruptions.
+    /// </para>
+    /// </remarks>
+    /// <seealso cref="GCHandle" />
+    /// <typeparam name="T">The type of the object this <see cref="GCHandle{T}"/> tracks to.</typeparam>
+    public struct GCHandle<T> : IEquatable<GCHandle<T>>, IDisposable
+        where T : class?
+    {
+        // The actual integer handle value that the EE uses internally.
+        private IntPtr _handle;
+
+        /// <summary>
+        /// Allocates a handle for the specified object.
+        /// </summary>
+        /// <param name="target">The object that uses the <see cref="GCHandle{T}"/>.</param>
+        public GCHandle(T target)
+        {
+            _handle = GCHandle.InternalAlloc(target, GCHandleType.Normal);
+        }
+
+        private GCHandle(IntPtr handle) => _handle = handle;
+
+        /// <summary>Determine whether this handle has been allocated or not.</summary>
+        public readonly bool IsAllocated => _handle != IntPtr.Zero;
+
+        /// <summary>Gets or sets the object this handle represents.</summary>
+        /// <exception cref="NullReferenceException">If the handle is not initialized or already disposed.</exception>
+        public readonly T Target
+        {
+            get
+            {
+                IntPtr handle = _handle;
+                GCHandle.CheckUninitialized(handle);
+                // Skip the type check to provide lowest overhead.
+                return Unsafe.As<T>(GCHandle.InternalGet(handle));
```

Comment:
This means if I do:
```C#
IntPtr p = GCHandle<A>.ToIntPtr(new GCHandle<A>(a));
B b = GCHandle<B>.FromIntPtr(p);
```
there's a type safety hole. I get that all of this is unsafe, but are we ok with that?

Reply:
I'm a bit concerned when trying to convert usages in libraries. There are some cases the type is asserted. Some assemblies are passing GCHandle of multiple types to unmanaged code, and there can be a chance to get things messed.